src/ngraph/runtime/gpu/gpu_shape.hpp

```diff
@@ -0,0 +1,119 @@
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include <cstdio>
+#include <stdexcept>
+#include <vector>
+
+#include "ngraph/axis_set.hpp"
+#include "ngraph/coordinate.hpp"
+#include "ngraph/shape.hpp"
+#include "ngraph/strides.hpp"
+
+namespace ngraph
+{
+    class Shape;
+    /// \brief Shape for a tensor resident on GPU.
+    class GPUShape : public std::vector<uint32_t>
+    {
+    public:
+        GPUShape(const std::initializer_list<uint32_t>& axis_lengths)
+            : std::vector<uint32_t>(axis_lengths)
+        {
+        }
+
+        GPUShape(const std::vector<uint32_t>& axis_lengths)
+            : std::vector<uint32_t>(axis_lengths)
+        {
+        }
+
+        GPUShape(const GPUShape& axis_lengths)
+            : std::vector<uint32_t>(axis_lengths)
+        {
+        }
+
+        explicit GPUShape(uint32_t n, uint32_t initial_value = 0)
+            : std::vector<uint32_t>(n, initial_value)
+        {
+        }
+
+        template <class InputIterator>
+        GPUShape(InputIterator first, InputIterator last)
+            : std::vector<uint32_t>(first, last)
+        {
+        }
+
+        GPUShape() {}
+        GPUShape& operator=(const GPUShape& v)
+        {
+            static_cast<std::vector<uint32_t>*>(this)->operator=(v);
+            return *this;
+        }
+
+        GPUShape& operator=(GPUShape&& v)
+        {
+            static_cast<std::vector<uint32_t>*>(this)->operator=(v);
+            return *this;
+        }
+
+        GPUShape(const Shape& shape)
+        {
+            for (size_t const& size : shape)
+            {
+                uint32_t low = static_cast<uint32_t>(size);
+                if (size >> 32 != 0)
+                {
+                    throw std::runtime_error(
+                        "Request for Shape which exceeds the bitwidth available for GPUShapes "
+                        "(32)");
+                }
+                this->push_back(low);
+            }
+        }
+
+        GPUShape(const Strides& strides)
+        {
+            for (size_t const& size : strides)
+            {
+                uint32_t low = static_cast<uint32_t>(size);
+                if (size >> 32 != 0)
+                {
+                    throw std::runtime_error(
+                        "Request for Strides which exceed the bitwidth available for GPUShapes "
+                        "(32)");
+                }
+                this->push_back(low);
+            }
+        }
+
+        GPUShape(const Coordinate& coord)
```

Comment:
32bit will definitely save space and computation. Do we have limitation on shape to 32bit? for example currently we treat all element-wised op as 1D, is it possible it will be more than 4G?

Reply:
There is no limitation on Shape, it can have dimension's of length >= 2^32. GPUShape removes support for shape sizes greater than 32 bits in the GPU transformer. If we want to support 64 bit descriptors in our kernels (CUDNN does not), we should do it as an option, but not by default because of the performance cost that you point out.

GPUShape makes the check that the shape is less than 32 bits in precision explicit. And it allows us to implicitly convert to 32 bit shapes in the transformer without needing to `static_cast` the high bits.

Reply:
>for example currently we treat all element-wised op as 1D, is it possible it will be more than 4G
Can you elaborate what you mean?

Reply:
I mean it's possible to exceed 32bit.

Reply:
GPUShape only needs to handle shapes that GPU kernels use. The 6d things from a framework will have been down-ranked by the time they get to GPUShape.

Reply:
Any operation shape that would be larger than the GPU can handle would first need to be transformed into a series of operations that fit, so encountering a GPUShape that was too big would be a bug.

Reply:
@diyessi @csullivan That's what I'm worried,  just an example, we have an 1K*1K*1K*1K 4D shape, when we transfer to 1D on GPU, it will be 1T.

Reply:
@diyessi GPU can handle 64bit address, here we are trying to see if we can manually limit it to 32bit to improve performance.

Reply:
For that, we'd want to first automatically chop the graph into pieces that fit in a way that wasn't too bad for performance (although any performance would be better than not being able to run at all). But first we want to handle the cases that do fit, since those are the kinds of graphs we would be chopping into.

Reply:
@fengleitian Shape{1000,1000,1000,1000} and GPUShape{1000,1000,1000,1000} are both allowed. You are right with the shape_size template, below, we should there return std::vector<size_t>, but this is the only problem I see. The only time GPUShape would limit us is if we had a GPUShape with one dimension greater than 32 bits in width.

The problem with what we are doing right now is that we pretend we support the full Shape, but in reality, most of our ops do not -- all CUDNN ops are limited to 2**32 elements per dimensions.